### ISSUE TITLE
Potential Ideas for Pneumaticcraft

### ISSUE THREAD

**Issue by Alice0Rose** [TeamPneumatic/pnc-repressurized#1112]
### Describe the feature

- Pressure Cutter (saw that chops down entire trees powered by pressurized air and water)
- Pneumatic Glove (melee weapon with high knockback, potentially chestplate upgrade?)
- Armor Upgrade Expander (adds more upgrade slots, applicable in pressure chamber)
- Pressurized Impact Reflector (late game armor upgrade that mitigates X% of damage receieved in exchange for proportional amount of air pressure)
- Solar Heat Generator (late game heat generator for use with thermal compressor)
- Manual Compressor (early game compressor powered by right clicks that takes more and more right clicks the higher the pressure, to simulate it getting harder to compressor the air)

### Reasons why it should be considered

- Nothing currently in mod like this, provides a means of manually gathering large amounts of wood for solid fuel compressors during the mid-game.
- Mod doesn't have any strong melee weapons currently, while having a slew of strong ranged weapons.
- Some armor pieces (such as helmet) have significantly more upgrades than the 9 slots allow, which limits the usability of some of the less powerful upgrades.
- Pneumatic armor is strong on its own, but pales in comparison to other mods. The late game of Pneumaticcraft is also somewhat lacking, so something like this could help to address both issues.
- A more realistic approach to solar power (instead of the place-and-forget style seen in other mods) could provide an interesting source of renewable energy without compromising the balanced nature of Pneumaticcraft. It would also add to the late game of the mod.
- An early source of pressurized air could allow a more diverse start to the mods progression instead of relying on solid fuel burning immediately.

### Additional details

I'm looking to implement these ideas myself via pull-requests, I just want to see which would be approved and which wouldn't before I start work on them.

**Comment by desht** [comment]
I actually quite like all of these, some comments though:
* Pneumatic Glove: yeah, the mod could do with a good melee weapon. You might also want to look at https://github.com/TeamPneumatic/pnc-repressurized/issues/729 (and check out the linked youtube video, it's pretty cool).  As for the knockback, it could make the Kick upgrade for the boots a bit redundant...
* Impact Reflector: the air cost should be significant, enough that an Aerial Interface and a lot of generation capability at your base is a requirement.
* Solar power: as long as there's some kind of balancing limitation to this, I could be OK with it (I don't like totally free energy sources; the Flux Compressor gets close to this depending on what RF mods you have, but at least there's the cooling requirement there).

**Comment by Alice0Rose** [comment]
- I thought about something like a bladed weapon, but couldn't think of a good way to have it fit into the theme of Pneumaticcraft. I'll do some research on it though and see if I can find something that fits. You're right about making the Kick upgrade redundant, so we can scrap the idea of it doing high knockback and just focus on damage output.
- My thoughts exactly. I was thinking you could configure it in the armor settings to mitigate X% of damage, with a nonlinear increase in air consumption the higher it goes.
- Ever since I first started modded Minecraft I've been against solar power because of how so many mods make it into this super cheap, free power that lasts forever and can power everything with next-to-no work. For Pneumaticcraft I was thinking that the solar panels would be have a max temperature, which after reaching they would stop working and have to be manually repaired. This would require managing the heat from the panels and ensuring they don't get too hot. However, as the panels get closer to their temperature maximum, they would become more efficient. This is the opposite of how they work in real life, but it would encourage players to put thought into their cooling to ensure it is _just_ enough to prevent overheat, but not so much that it kills the panels' efficiency. I was also thinking of adding the option to close the panels via redstone/ComputerCraft, so players could setup a failsafe of sorts, or only have the panels active during certain times of the day to prevent overheat, etc. TLDR, players will have to put thought and design into their solar panels if they want them to be more than expensive paperweights.

**Comment by Alice0Rose** [comment]
Just some polishing on some of the ideas I posted before along with some new ones I've been mulling over.

- Pressure Cutter: mid game, saw that chops down entire trees powered by pressurized air and water
- Pneumatic Gauntlet: mid game, melee weapon, applies slowness on hit
- Ionized Air Blade: late game, melee weapon, deals some armor piercing damage along with igniting on hit
- Armor Upgrade Expander: late game, adds more upgrade slots to armor, applicable in pressure chamber
- Pressurized Impact Reflector: late game, armor upgrade that mitigates X% of damage receieved in exchange for proportional amount of air pressure
- Solar Compressor: late game, compressor that generates air and heat from sunlight
- Manual Compressor: early game, compressor powered by right clicks that takes more and more right clicks the higher the pressure
- Ore Processing: compress 2 raw ore in pressure chamber into compressed ore (2x), refine in thermopneumatic processing plant with water (3x) or lubricant (4x), smelt results for ingots
- NPC Hacking Bases: late game, areas with large amounts of loot that require hacking 1+ security stations, along with dealing with turrets and drones, potentially unique loot needed for late game tools (Air Blade/Pressurized Impact Reflector)

Also almost done with the Solar Compressor, just need to fine-tune the air to heat generation ratio so keeping it at a balanced heat level is feasible, even with speed upgrades. Been sick the past week or so, so haven't had the chance to work on it much.

**Comment by M3wP** [comment]
The melee weapon would have to be in the style of War Hammer...  Or perhaps The Evil Dead..  Think part sword, part chainsaw?  Perhaps you could merge a few of those into one.

**Comment by desht** [comment]
* Regarding the Pressure Cutter: the Jackhammer with a diamond bit pretty much fills that role since it can do veinmining for ores and logs.  And the diamond bit is mid-game (needs plastic, but can be done before tier 2 pressure).
* Regarding an ore processing system - HipHopHuman did a very comprehensive design here: https://github.com/TeamPneumatic/pnc-repressurized/issues/778#issuecomment-926586607 Not saying that's definitely the way to go, but there are some interesting ideas there for sure...
* Regarding NPC hacking - cool idea. Something that I've also toyed with is the idea of hostile rogue drone mobs, probably limited to spawning in specific structures. Could be given random weapons, either melee or minigun-upgrade, or even micromissiles (ouch).  And drop random useful loot, like components for PCB and drone manufacture (actual working Finished PCB's would be v. rare).

**Comment by Alice0Rose** [comment]
- I didn't realize the Jackhammer could be used for cutting down trees; I always only used it as a mining tool. Something I noticed while testing it out though is that the Diamond Bit's vein mining isn't limited to 9 blocks, despite what the tooltip on it says. If it was, I could still see the Pressure Cutter (now named the Water Jet Cutter, after I've done some research on it and started modelling it after the real-life equivalent) having a place, being cheaper than the Jackhammer and able to fell large trees (including their leaves), which would normally require the Netherite Bit. So it depends on if you want to limit the Diamond Bit's vein mining to match its tooltip or not. If not, I'll shelf the idea for now and come back to it later if I can think of a way to make the Water Jet Cutter unique.
- Oh wow, yeah that's super in-depth, I'll definitely look through that once I get to that feature.
- That's exactly what I was thinking, and then if you hacked all the security stations of the base, it would either destroy all the drones, or make them friendly, allowing you to pick them up.

**Comment by desht** [comment]
That tooltip on the Diamond Bit is actually misleading and something I need to fix.  It shows the blocks dug by the "best" dig type available for the bit type, which for diamond is the full 3x3 mode.  When in ore/logs veinmining mode, it's 128 blocks.

Problem is "best" is rather situational, so I'll either just remove that part of the tooltip, or see if it can be made sensitive to the current dig type rather than installed drill bit.

**Comment by Alice0Rose** [comment]
I see, that makes sense. So it's intentional that the diamond drill bit vein mining has a range of 128 blocks? I'm just thinking that might be a bit strong, not necessarily for wood, but for some of the massive ore veins that were added in 1.18.

**Comment by desht** [comment]
Yeah the Jackhammer was added back in 1.16, but does vanilla actually add huge ore veins like that these days?  I haven't come across any. Also bear in mind that pressure is used for each block mined, so mining veins like that isn't entirely free.

**Comment by Alice0Rose** [comment]
This is what I found about it officially: https://minecraft.fandom.com/wiki/Ore_vein.

My concern is more with the Diamond Bit compared to the Netherite Bit. If the only difference between the two is Veinminer+, the cost difference seems disproportional given the rarity of netherite and requiring PCBs and the Assembly Controller.

**Comment by desht** [comment]
Netherite is a bit faster as well, but yeah it's not a huge upgrade over diamond at this point.  I have been considering boosting the speed of Netherite bits significantly (without any air cost increase) to compensate.

Given ore veins are meant to be rare, I'm not too bothered about allowing lots of blocks.  Maybe reduce diamond to 32 or 64, and allow 128 for netherite.

**Comment by HipHopHuman** [comment]
Just a note on my ore multiplication design, it was designed before the modded ecosystem caught up to the way ores now work in the newer MC versions, so if you're planning on implementing it any close to the way I designed it, just bare that in mind :)

**Comment by Alice0Rose** [comment]
> Netherite is a bit faster as well, but yeah it's not a huge upgrade over diamond at this point. I have been considering boosting the speed of Netherite bits significantly (without any air cost increase) to compensate.
> 
> Given ore veins are meant to be rare, I'm not too bothered about allowing lots of blocks. Maybe reduce diamond to 32 or 64, and allow 128 for netherite.

That makes sense to me. So what are we thinking for the Water Jet Cutter? The model for it is done, but I haven't coded the item yet. Should I shelve it for now, or move forward with it?

**Comment by Alice0Rose** [comment]
> Just a note on my ore multiplication design, it was designed before the modded ecosystem caught up to the way ores now work in the newer MC versions, so if you're planning on implementing it any close to the way I designed it, just bare that in mind :)

Yeah we'd probably want to change the actual raw ore -> ingot rates to something similar to recent mekanism versions, since they account for the ability to fortune all ores now, but the processes you suggested can probably remain the same.

**Comment by johalun** [comment]
> hostile rogue drone mobs

YES YES YES! :D
